### PR TITLE
Bear.py: Add attrs to Bears v2

### PR DESCRIPTION
**Pull request by Siddhartha1234**
Add attrs for section, file_dict and _dependency_results

Closes https://github.com/coala/coala/issues/3864

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

**Comment by ashwini0529** [comment]
@Siddhartha1234 try always to run `coala` on your code before pushing it. 😄

File: coalib/core/Bear.py [hunk 1]
```diff
@@ -179,25 +185,11 @@ def maintainers_emails(cls):
                 if cls.MAINTAINERS_EMAILS == set() else
                 cls.MAINTAINERS_EMAILS)
 
-    @enforce_signature
-    def __init__(self, section: Section, file_dict: dict):
+    def __attrs_post_init__(self):
```
Comment (adhikasp) [inline coalib/core/Bear.py]: Does this mean the rather helpful docs string on how to construct a bear and it's argument get lost? Is there anyway to keep the doc string?

Comment (Siddhartha1234) [inline coalib/core/Bear.py]: @adhikasp I can keep the Docstring If you wish . Any tip @Makman2 @Udayan12167 ?

Comment (Makman2) [inline coalib/core/Bear.py]: Definitely don't delete the docstring

Comment (Makman2) [inline coalib/core/Bear.py]: Or try to incorporate it into the auto-generated constructor.

File: coalib/core/Bear.py [hunk 2]
```diff
@@ -179,23 +196,24 @@ def maintainers_emails(cls):
...
+        """
+        This function is called after all attributes are initialized.
+        It checks for pre-requisites.
         """
-        self.section = section
-        self.file_dict = file_dict
 
-        self._dependency_results = defaultdict(list)
+        if not isinstance(self.section, Section):
```
Comment (Siddhartha1234) [inline coalib/core/Bear.py]: added manual isinstance check as enforce_signature is not allowed for Attributes and tests were there to assert TypeErrors. This can be improved by including enforce_signature support for Attributes in coala-utils. @Makman2  is it okay for now?

File: requirements.txt [hunk 3]
```diff
@@ -8,3 +8,5 @@ Pygments~=2.1
 PyPrint~=0.2.6
 requests~=2.12
 setuptools>=17.0
+attrs~=16.3.0
+
```
Comment (Makman2) [inline requirements.txt]: redundant newline

File: coalib/core/Bear.py [hunk 4]
```diff
@@ -130,6 +132,21 @@ class Bear:
...
+        Attributes for a new bear.
+
+        :param section:
+            The section object where bear settings are contained.
+        :param file_dict:
+            The file-dictionary containing a mapping of filenames to the
+            according file contents.
+        :raises RuntimeError:
+            Raised when bear requirements are not fulfilled.
+    """
```
Comment (Makman2) [inline coalib/core/Bear.py]: Hm I want to have this documentation in the constructor somehow... or somewhere available where everyone would look instantly. Is there some possibility? @Siddhartha1234

File: coalib/core/Bear.py [hunk 5]
```diff
@@ -130,6 +132,21 @@ class Bear:
...
+    section = attr.ib(hash=False)
+    file_dict = attr.ib(hash=False)
+    _dependency_results = attr.ib(init=False, default=defaultdict(list),
+                                  hash=False)
```
Comment (Makman2) [inline coalib/core/Bear.py]: actually where's the difference over assigning those inside a custom `__init__`?

File: requirements.txt (hunk 3 above)
Comment (Siddhartha1234) [inline requirements.txt]: Travis was not recognizing attrs~=16.3.0 unless there was  a newline .It was saying module not found . Other CIs had no issues.

File: coalib/core/Bear.py (hunk 5 above)
Comment (Siddhartha1234) [inline coalib/core/Bear.py]: @Makman2 attr is a decorator which generates an __init__ flexibly . The whole point is not only about not having to write an init. We can utilize the properties of attrs .Like , we can obtain all attributes as a dict or json .We can attach hooks to each attribute which will execute after that is initialized etc. That should be a separate PR IMO . I'll be glad to add that in a new PR. This PR just adds the attrs, not It's usecases :) .

File: requirements.txt (hunk 3 above)
Comment (Makman2) [inline requirements.txt]: that sounds really strange and this shouldn't happen. Please remove this line. If CI fails I'll look at it :+1:

File: coalib/core/Bear.py (hunk 5 above)
Comment (Makman2) [inline coalib/core/Bear.py]: > we can obtain all attributes as a dict or json

this is in fact already possible by using `vars(instance)` or `dir(instance)` ;)

> We can attach hooks to each attribute which will execute after an attribute is initialized 

so that's what we would do inside the `__init__` anyway, right?

> This PR just adds the attrs, not It's usecases :)

Yeah I understood that intention already from @sils :) But the bear is in fact working, so I don't really see to change something in terms of "layout".

Don't want to sound too strict or so, I just want to settle things with `attrs`, as it came accross me from some people, telling me like "oh man super cool stuff", but I can't see that for `coalib.core.Bear` :/

Comment (all3fox) [inline coalib/core/Bear.py]: @Makman2 just two cents here: `attrs` generates comparisons/string representations "quickly" and only gives slight overhead at class definition but not at runtime. These might be the "selling points" you are trying to hear.

As for this particular use case, `attrs` probably does not differ much from just writing the `__init__` manually. You could argue, that you do not have to implement `__repr__` and `__str__` and can even compare bears with `__le__` and the like.

Comment (Makman2) [inline coalib/core/Bear.py]: for `repr` we have thankfully `@generate_repr` :D

> ... and can even compare bears with __le__ and the like.

The problem is: it makes no sense to compare them, except `==` (though we need to be also careful with that, the new core defines whether considering two bears equal, this plays a large role in the bear-dependency system). The default instance check is the safest currently imo^^

Many other classes would really benefit from `attr.s`, though it's a bit tricky for this one. It's used in multiple processes and contains partially unhashable content^^

Comment (Siddhartha1234) [inline coalib/core/Bear.py]: @Makman2 As Lasse said , even though the usecases might be limited or tricky as you mentioned.It's a good idea to store all data members which are holding important data as an Attribute. So that the `attrs.as_dict` of the entire core repo will contain the important data members.This can be a starting point to improve our Data Model. Distinguishing between data_members is the most important aspect of `attrs` , other features have been implemented in some way or the other in the repo already.

File: coalib/core/Bear.py [hunk 6]
```diff
@@ -179,23 +196,24 @@ def maintainers_emails(cls):
...
+        """
+        This function is called after all attributes are initialized.
+        It checks for pre-requisites.
         """
-        self.section = section
-        self.file_dict = file_dict
 
-        self._dependency_results = defaultdict(list)
+        if not isinstance(self.section, Section):
+            raise TypeError('section must be a Section')
```
Comment (sils) [inline coalib/core/Bear.py]: you can add a validator in attrs

File: coalib/core/Bear.py (hunk 5 above)
Comment (sils) [inline coalib/core/Bear.py]: FWIW I really like this declarative style vs. having a custom init

File: coalib/core/Bear.py (hunk 6 above)
Comment (Siddhartha1234) [inline coalib/core/Bear.py]: @sils That way i have to define 3 different functions outside the class having the same content as these 3 lines separately as validators don't auto-raise exceptions and just accept a callable entity. Is that okay ?

File: coalib/core/Bear.py [hunk 7]
```diff
@@ -179,24 +198,19 @@ def maintainers_emails(cls):
                 if cls.MAINTAINERS_EMAILS == set() else
                 cls.MAINTAINERS_EMAILS)
 
-    @enforce_signature
-    def __init__(self, section: Section, file_dict: dict):
+    @staticmethod
+    def setup_dependencies():
```
Comment (Makman2) [inline coalib/core/Bear.py]: hm where does this one come from?

File: coalib/core/Bear.py (hunk 5 above)
Comment (Makman2) [inline coalib/core/Bear.py]: Still not sure, though let's do it and see how it turns out :+1:

File: coalib/core/Bear.py [hunk 8]
```diff
@@ -130,6 +132,23 @@ class Bear:
...
+    """
+        Attributes for a new bear.
+
+        :param section:
+            The section object where bear settings are contained.
+        :param file_dict:
+            The file-dictionary containing a mapping of filenames to the
+            according file contents.
+        :raises RuntimeError:
+            Raised when bear requirements are not fulfilled.
```
Comment (Makman2) [inline coalib/core/Bear.py]: as we can't document `__init__` any more, we should move that to class level and embed it into the text somehow, like
```
On instantiation, a `RuntimeError` is raised when the bear requirements are not fulfilled.
```

File: coalib/core/Bear.py [hunk 9]
```diff
@@ -179,24 +198,19 @@ def maintainers_emails(cls):
...
-        self.section = section
-        self.file_dict = file_dict
-
-        self._dependency_results = defaultdict(list)
 
+    def __attrs_post_init__(self):
+        """
+        This function is called after all attributes are initialized.
+        It checks for pre-requisites.
+        """
```
Comment (Makman2) [inline coalib/core/Bear.py]: As `__attrs_post_init__` is known how to work for `attrs`, you could omit the docstring. But up to you :)

File: requirements.txt [hunk 10]
```diff
@@ -8,3 +8,4 @@ Pygments~=2.1
 PyPrint~=0.2.6
 requests~=2.12
 setuptools>=17.0
+attrs~=16.3.0
```
Comment (Makman2) [inline requirements.txt]: could you place it in the alphabetically right position? makes it easier to find ;)

File: coalib/core/Bear.py (hunk 7 above)
Comment (Siddhartha1234) [inline coalib/core/Bear.py]: I misplaced that function while deleting init, Will fix it in next update.

File: coalib/core/Bear.py (hunk 8 above)
Comment (Siddhartha1234) [inline coalib/core/Bear.py]: @Makman2 You mean in the main docstring? Or in a class property ?

File: coalib/core/Bear.py (hunk 9 above)
Comment (Siddhartha1234) [inline coalib/core/Bear.py]: Okay, I will . :)

File: coalib/core/Bear.py (hunk 8 above)
Comment (Makman2) [inline coalib/core/Bear.py]: in the docstring of the class :+1:

File: coalib/core/Bear.py [hunk 11]
```diff
@@ -28,6 +30,15 @@ class Bear:
     will probably want to look at the ``ProjectBear`` and ``FileBear`` classes
     that inherit from this class.
 
+    Attributes for a new bear:
+
+    section:   The section object where bear settings are contained.
+    file_dict: The file-dictionary containing a mapping of filenames to the
+    according file contents.
+
+    On instantiation, a `RuntimeError` is raised when the bear requirements
```
Comment (Makman2) [inline coalib/core/Bear.py]: `RuntimeError` in double backticks :+1:

File: coalib/core/Bear.py [hunk 12]
```diff
@@ -28,6 +30,15 @@ class Bear:
     will probably want to look at the ``ProjectBear`` and ``FileBear`` classes
     that inherit from this class.
 
+    Attributes for a new bear:
```
Comment (Makman2) [inline coalib/core/Bear.py]: Maybe you should incorporate a doctest, something like
```
To instantiate a bear, pass a section which contains the bear setting and a file-dictionary (that maps filenames to the according file contents):

>>> section = Section('my-section')
>>> file_dict = {'file.py': ['This is a one-liner\n'']}
>>> Bear(section, file_dict)
```
As when I read "Attributes for a new bear", I wouldn't relate that I have to pass the following values into the constructor.

File: requirements.txt [hunk 13]
```diff
@@ -7,4 +8,4 @@ libclang-py3~=3.4.0
 Pygments~=2.1
 PyPrint~=0.2.6
 requests~=2.12
-setuptools>=17.0
+setuptools>=17.0
```
Comment (Makman2) [inline requirements.txt]: hm this change doesn't seem to change anything

File: coalib/core/Bear.py [hunk 14]
```diff
@@ -179,24 +197,11 @@ def maintainers_emails(cls):
...
-
-        :param section:
-            The section object where bear settings are contained.
-        :param file_dict:
-            The file-dictionary containing a mapping of filenames to the
-            according file contents.
-        :raises RuntimeError:
-            Raised when bear requirements are not fulfilled.
+        This function is called after all attributes are initialized.
+        It checks for pre-requisites.
```
Comment (Makman2) [inline coalib/core/Bear.py]: "prerequisite" is a legal word, so I would remove the `-` if you want to keep the docstring ;)

File: requirements.txt (hunk 13 above)
Comment (Makman2) [inline requirements.txt]: Please put this one into a new commit with a nice commit message containing that it screws up your CI ;)

File: coalib/core/Bear.py [hunk 15]
```diff
@@ -28,6 +29,16 @@ class Bear:
     will probably want to look at the ``ProjectBear`` and ``FileBear`` classes
     that inherit from this class.
 
+    To instantiate a bear, pass a section which contains the bear setting and
```
Comment (adtac) [inline coalib/core/Bear.py]: IMO this should be in its own commit. A commit should do exactly one thing.

Comment (adtac) [inline coalib/core/Bear.py]: I fully welcome the addition :)

File: coalib/core/Bear.py (hunk 1 above)
Comment (adtac) [inline coalib/core/Bear.py]: Was the docstring somewhere else? If so, ignore my previous comment.

File: requirements.txt (hunk 13 above)
Comment (adtac) [inline requirements.txt]: what's the change here?

File: coalib/core/Bear.py (hunk 15 above)
Comment (Makman2) [inline coalib/core/Bear.py]: Imo it's okay, as the constructor docstring is moved into the class docstring.

File: coalib/core/Bear.py [hunk 16]
```diff
@@ -131,6 +142,13 @@ class Bear:
     ASCIINEMA_URL = ''
     BEAR_DEPS = set()
 
+    section = attr.ib(hash=False,
+                      validator=attr.validators.instance_of(Section))
+    file_dict = attr.ib(hash=False,
+                        validator=attr.validators.instance_of(dict))
+    _dependency_results = attr.ib(init=False, default=defaultdict(list),
```
Comment (sils) [inline coalib/core/Bear.py]: careful this is mutable

Comment (sils) [inline coalib/core/Bear.py]: use attr.Factory as per https://attrs.readthedocs.io/en/stable/examples.html#defaults

**Comment by sils** [comment]
@Siddhartha1234 let us know if we should take this over, it's almost ready for merge and just needs small changes, a rebase and the splitting of the commits

**Comment by Makman2** [comment]
Closing due to inactivity.
